### PR TITLE
[API Compatiblity] add compat.softmax to compat with torch.softmax

### PR DESCRIPTION
**Pull request by cyy536**
### PR Category
User Experience

### PR Types
New features

### Description
add compat.softmax to compat with torch.softmax

### CODE REVIEW

**Review by zhwesky2010 — COMMENTED** [review]
这里为啥还分了这么多softmax。

只有两个softmax，compat.softmax将dim进行设置直接调用原来的softmax。

**Comment by cyy536** [comment]
一共写了4个，它们类型不一样
paddle.Tensor.softmax==torch.Tensor.softmax的类型是 (int dim, torch.dtype dtype = None)
paddle.softmax==torch.softmax的类型是 (Tensor input, int dim, torch.dtype dtype = None, *, Tensor out = None)
paddle.compat==torch.nn.functional.softmax的类型是 (input, dim=None, _stacklevel=3, dtype=None)
paddle.nn.functional.softmax的类型是 (input, dim=None, _stacklevel=3, dtype=None)，但默认值不一样

File: python/paddle/tensor/softmax.py [hunk 1]
```diff
@@ -159,7 +164,156 @@ def softmax(
               [0.03205860, 0.08714432, 0.23688282, 0.64391426],
               [0.03205860, 0.08714432, 0.23688282, 0.64391426]]])
     """
+    return _softmax_impl(x, axis, dtype, name, out=out)
+
+
+@softmax_param_ignore_alias
+def compat_softmax(
```
Comment (zhwesky2010) [inline python/paddle/tensor/softmax.py]: 这个实现放到对应位置里去吧，`tensor/compat.py` 里。`paddle.compat.softmax` 可以直接调 `paddle.softmax`

File: python/paddle/tensor/softmax.py [hunk 2]
```diff
@@ -159,7 +164,156 @@ def softmax(
               [0.03205860, 0.08714432, 0.23688282, 0.64391426],
               [0.03205860, 0.08714432, 0.23688282, 0.64391426]]])
     """
+    return _softmax_impl(x, axis, dtype, name, out=out)
```
Comment (zhwesky2010) [inline python/paddle/tensor/softmax.py]: 这个可以直接展开

File: python/paddle/tensor/softmax.py [hunk 3]
```diff
@@ -159,7 +164,156 @@ def softmax(
...
+
+
+@softmax_param_ignore_alias
+def compat_softmax(
+    x: Tensor,
```
Comment (zhwesky2010) [inline python/paddle/tensor/softmax.py]: compat下的签名风格就按 input/dim 来吧，与`min/max/sort` 这些保持一致。

File: python/paddle/compat.py [hunk 4]
```diff
@@ -19,5 +19,6 @@
     sort,
     split,
 )
+from .tensor.softmax import compat_softmax as softmax
```
Comment (zhwesky2010) [inline python/paddle/compat.py]: 去 .tensor.compat 下实现。

File: python/paddle/tensor/softmax.py [hunk 5]
```diff
@@ -159,7 +164,156 @@ def softmax(
               [0.03205860, 0.08714432, 0.23688282, 0.64391426],
               [0.03205860, 0.08714432, 0.23688282, 0.64391426]]])
     """
+    return _softmax_impl(x, axis, dtype, name, out=out)
+
+
+@softmax_param_ignore_alias
```
Comment (zhwesky2010) [inline python/paddle/tensor/softmax.py]: 这个能复用当前的一些基建吗，保持风格一致的warning和提示比较好。参考 `paddle.compat.min` 风格。

infoflow 2025-08-26 18-17-31

File: python/paddle/tensor/softmax.py (hunk 1 above)
Comment (cyy536) [reply python/paddle/tensor/softmax.py]: 已调整

File: python/paddle/tensor/softmax.py (hunk 2 above)
Comment (cyy536) [reply python/paddle/tensor/softmax.py]: 已展开

File: python/paddle/tensor/softmax.py (hunk 3 above)
Comment (cyy536) [reply python/paddle/tensor/softmax.py]: 已调整

File: python/paddle/tensor/softmax.py (hunk 5 above)
Comment (cyy536) [reply python/paddle/tensor/softmax.py]: 已复用

File: python/paddle/tensor/__init__.py [hunk 6]
```diff
@@ -494,6 +494,8 @@
 )
 from .to_string import set_printoptions  # noqa: F401
 
+__all__ = ['softmax']
```
Comment (zhwesky2010) [inline python/paddle/tensor/__init__.py]: 这个不能放__all__里，这个要放到下面的tensor_method_func，才会bind到paddle.Tensor上

File: python/paddle/tensor/compat.py [hunk 7]
```diff
@@ -830,3 +836,199 @@ def max(
...
+    if (
+        (dtype is not None)
+        and (not isinstance(dtype, core.VarDesc.VarType))
+        and (not isinstance(dtype, core.DataType))
+    ):
+        dtype = convert_np_dtype_to_dtype_(dtype)
+    if in_dynamic_or_pir_mode():
+        outs_cast = input if dtype is None else _C_ops.cast(input, dtype)
+        return paddle.assign(_C_ops.softmax(outs_cast, dim), out)
+    else:
```
Comment (zhwesky2010) [inline python/paddle/tensor/compat.py]: 下面的这个老静态图的的分支不需要实现

File: python/paddle/tensor/compat.py [hunk 8]
```diff
@@ -29,11 +30,16 @@
...
     from paddle._typing import (
+        DTypeLike,
         Size2,
     )
 
 
 from paddle import nn
+from paddle.base.data_feeder import check_dtype, check_variable_and_dtype
+from paddle.base.framework import convert_np_dtype_to_dtype_
+from paddle.base.layer_helper import LayerHelper
```
Comment (zhwesky2010) [inline python/paddle/tensor/compat.py]: 老静态图分支不需要实现，移除无需导入的东西

File: python/paddle/tensor/compat.py [hunk 9]
```diff
@@ -830,3 +836,199 @@ def max(
...
+
+@ForbidKeywordsDecorator(
+    illegal_keys={"x", "axis", "name"},
+    func_name="paddle.compat.softmax",
+    correct_name="paddle.nn.functional.softmax",
+)
+def softmax(
+    input: Tensor,
+    dim: int | None = None,
+    _stacklevel: int = 3,
```
Comment (zhwesky2010) [inline python/paddle/tensor/compat.py]: 这个能对齐到torch.softmax吗，这个_stacklevel参数

你是不是继承下ForbidKeywordsDecorator，隐藏处理下_stacklevel，然后外面的签名还是按`input, dim, dtype, *, out`来

File: python/paddle/tensor/compat.py [hunk 10]
```diff
@@ -830,3 +836,199 @@ def max(
...
+            dim = 1
+    if (
+        (dtype is not None)
+        and (not isinstance(dtype, core.VarDesc.VarType))
+        and (not isinstance(dtype, core.DataType))
+    ):
+        dtype = convert_np_dtype_to_dtype_(dtype)
+    if in_dynamic_or_pir_mode():
+        outs_cast = input if dtype is None else _C_ops.cast(input, dtype)
+        return paddle.assign(_C_ops.softmax(outs_cast, dim), out)
```
Comment (zhwesky2010) [inline python/paddle/tensor/compat.py]: 这个直接 _C_ops.softmax(outs_cast, dim, out=out)

File: python/paddle/tensor/softmax.py [hunk 11]
```diff
@@ -121,6 +121,7 @@ def softmax(
             :math:`axis + D` . Default is -1.
         dtype (str, optional): The data type of the output tensor, can be bfloat16, float16, float32, float64.
         name (str|None, optional): For details, please refer to :ref:`api_guide_Name`. Generally, no setting is required. Default: None.
+        out (Tensor, optional): The output Tensor.
```
Comment (zhwesky2010) [inline python/paddle/tensor/softmax.py:125]: 这个能用 _C_ops.softmax(outs_cast, axis, out=out) 来实现吗

File: python/paddle/tensor/__init__.py [hunk 12]
```diff
@@ -31,6 +31,7 @@
     real,
     shape,
 )
+from .compat_softmax import softmax as softmax
```
Comment (zhwesky2010) [inline python/paddle/tensor/__init__.py:34]: 这个得加到tensor_method_func，才会bind到paddle.Tensor上

File: python/paddle/utils/decorator_utils.py [hunk 13]
```diff
@@ -403,6 +396,70 @@ def process(
         return args, kwargs
 
 
+class ForbidKeywordsIgnoreOneParamDecorator(DecoratorBase):
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这个可以继承ForbidKeywordsDecorator，只抽出它的ignore逻辑，其他采用 `super().init(*args, **kwargs)` ，这样可以最大化复用代码。

File: python/paddle/tensor/__init__.py (hunk 12 above)
Comment (cyy536) [reply python/paddle/tensor/__init__.py:34]: tensor_method_func一直都有加

File: python/paddle/utils/decorator_utils.py (hunk 13 above)
Comment (cyy536) [reply python/paddle/utils/decorator_utils.py]: 已修改

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by SigureMo — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
